### ISSUE TITLE
Using dropout layers

### ISSUE THREAD

**Issue by Jendker** [davisking/dlib#1631]
I would like to ask what is the proper way to use the dropout layers, I didn't find any suitable example which would describe it.

If we are about to use dropout layers should we use the dropout layers just for training and then replace them with affine transforms during testing similarly as it is done with batch normalization layers? Or is there something more we should in testing to account for earlier removed neurons in the training when the dropout was active?

**Comment by headupinclouds** [comment]
You can use the same pattern that you use with batch normalization + affine.  To make inference reproducible, you can just replace the `dlib::dropout` layer with a `dlib::multiply` layer.   You can do this by just deserializing your training network (with dropout) into a matching inference network where the `dlib::multiply` layer occurs at the same location.  They share a default constructor value of 0.5.  If you use a different value, you will need to set it explicitly for the node at runtime. 

Something like:
```
dlib::layer<2>(net).layer_details()   = dlib::dropout_(dropout_rate);
```

**Comment by Jendker** [comment]
Now I found the post here... https://sourceforge.net/p/dclib/discussion/442518/thread/b366359f3b/?limit=25#388f/195d So I just replace `dlib::dropout` with `dlib::multiply`, deserialise the model file and I am good to go, without any values adjustment.

I'll give it a try, thanks!

**Comment by Jendker** [comment]
Found in the resources: http://dlib.net/dlib/dnn/layers_abstract.h.html that one can even set value of `dlib::multiply` with `dlib::dropout_(dropout_rate)`, so this must be the correct way of doing it. Thank you!
